### ISSUE TITLE
copilot AI unaware of Angular 17, or anything past Angular 12

### ISSUE THREAD

**Issue by godind** [microsoft/vscode-copilot-release#1128]
### godind: this project uses angular 17

GitHub Copilot: I apologize for the confusion earlier. As an AI, I'm based on the latest available information, and **as of my last update, the latest stable version of Angular is Angular 12**. If Angular 17 does indeed support a `@for` loop syntax in its templates, I'm not currently aware of it.

However, if you're experiencing issues with your current code, it might be due to the `track` keyword. Angular uses `trackBy` in `*ngFor` loops to improve performance. If Angular 17's `@for` loop follows a similar pattern, you might need to change `track` to `trackBy`.

Here's how your code might look:

```html
<div [formGroup]="multiCtrlArray">
  @for(ctrl of multiCtrlArray.controls; trackBy ctrl; let i = $index) {
    <boolean-control-config class="full-width"
      [ctrlFormGroup]="getFormGroup(ctrl)"
      [controlIndex]="i"
      [arrayLength]="arrayLength"
      (moveUp)="moveUp($event)"
      (moveDown)="moveDown($event)"
      (deleteCtrl)="deletePath($event)">
    </boolean-control-config>
  }
  <button class="add-btn" type="button" mat-mini-fab color="primary" (click)="addCtrlGroup()" aria-label="Add control down in the list">
    <i class="fa-solid fa-plus fa-xl"></i>
  </button>
</div>
```

If this doesn't solve your problem, I recommend checking the Angular 17 documentation or reaching out to the Angular community for more specific help.

### godind: How can I report your problem of not knowing angular 17 and @for and @if directives

GitHub Copilot: I apologize for any inconvenience caused. As an AI developed by OpenAI, I'm trained on a diverse range of data sources, including many programming languages and frameworks. **However, my training data only includes information available up to September 2021, and as of that time, the latest stable version of Angular was Angular 12.**

If there have been updates or changes to Angular after this time, including a new version (Angular 17) or new directives (@for, @if), I wouldn't be aware of them. 

To report this issue, you can provide feedback directly to OpenAI. They continuously work on updating and improving the models based on user feedback. 

Again, I apologize for any confusion or inconvenience. Your understanding is appreciated.

**Comment by IAbhickleton** [comment]
This is getting a bit ridiculous to the point of being unusable with Angular. Are there any updates on this?

**Comment by godind** [comment]
Agreed! It make copilot misleading with anything hint past angular 16 and getting worse.

**Comment by digitarald** [comment]
The latest model, GPT-4o can handle the Angular 16. Are there examples for the latest Copilot Chat version of Angular 17?

**Comment by IAbhickleton** [comment]
@digitarald It definitely improved with 4o. It's now aware of signals but constantly corrects my usage of the new control flow syntax in template files. This is an Angular 17 feature (we're now on 18).

**Comment by godind** [comment]
I agree v18 template flow need to be added as all current AI suggestions is suggesting to
remove performance improvements and simplified code.

This version is better but not fully there yet.

**Comment by levinzz** [comment]
> The latest model, GPT-4o can handle the Angular 16. Are there examples for the latest Copilot Chat version of Angular 17 not working?

**Me** : _adjust all cases in the selected code to use the new case syntax_

**Github copilot (GPT 4o)** : _To adjust the code to use the new Angular control flow syntax, you can replace the @switch and @case directives with Angular's *ngSwitch and *ngSwitchCase directives._ 

 ... _produces wrong code_ ...

_By using the new *ngSwitch and *ngSwitchCase directives, you can make your templates more concise and readable, leveraging Angular's latest features for control flow._

**Comment by godind** [comment]
What is this taking so long? It the primary reason I pay for copilot!!!!

**Comment by godind** [comment]
Hi. I am still paying every month to get AI on anguler. It's not up to day and way beyond. It keeps correcting perfectly good code and not understanding new angular features. PLEASE FIX OR I WILL CANCEL MY MONTHLY.

Thank you

**Comment by dmitrivMS** [comment]
We have just released a major update to Copilot functionality in VS Code (https://code.visualstudio.com/updates/v1_107) which includes many bug fixes as well. Could you please try the latest version to see if the problem still occurs? If it does, please open a new issue in the https://github.com/microsoft/vscode repo.

Have you tried supplying context for the framework you need through something like https://context7.com/?q=angular?
